### ISSUE TITLE
Skydns throws IO errors on large cluster, slow query performance.

### ISSUE THREAD

**Issue by fluxrad** [kubernetes/kubernetes#19634]
A prod cluster consisting of 40 nodes and ~1100 pods seems to be bottlenecked by DNS queries. Looking at the skydns container logs I see lots of io errors talking to etcd:

```
2016/01/14 02:09:45 skydns: error from backend: 501: All the given peers are not reachable (failed to propose on members [http://localhost:4001] twice [last error: Get http://localhost:4001/v2/keys/skydns/local/cluster/svc/default/com/foo/bar?quorum=false&recursive=true&sorted=false: dial tcp 127.0.0.1:4001: i/o timeout]) [0]
2016/01/14 02:09:45 skydns: error from backend: 501: All the given peers are not reachable (failed to propose on members [http://localhost:4001] twice [last error: Get http://localhost:4001/v2/keys/skydns/local/cluster/svc/default/com/foo/bar?quorum=false&recursive=true&sorted=false: dial tcp 127.0.0.1:4001: i/o timeout]) [0]
2016/01/14 02:09:45 skydns: error from backend: 501: All the given peers are not reachable (failed to propose on members [http://localhost:4001] twice [last error: Get http://localhost:4001/v2/keys/skydns/local/cluster/svc/default/com/foo/bar?quorum=false&recursive=true&sorted=false: dial tcp 127.0.0.1:4001: i/o timeout]) [0]
2016/01/14 02:31:14 skydns: failure to return reply "write udp: invalid argument"
2016/01/14 02:31:31 skydns: failure to return reply "write udp: invalid argument"
2016/01/14 02:31:45 skydns: failure to return reply "write udp: invalid argument"
2016/01/14 02:33:16 skydns: failure to return reply "write udp: invalid argument"
2016/01/14 02:33:21 skydns: failure to return reply "write udp: invalid argument"
```

We see that pods are — in `/etc/resolv.conf` — being given `ndots:5` which would, if I'm not mistaken, generate five queries per external lookup (one for each entry in the search path, and a final, recursive query)? The [large cluster](http://kubernetes.io/v1.1/docs/admin/cluster-large.html) docs indicate that scaling the size of the kube-dns pod is indicated here. I've confirmed that scaling out multiple kube-dns pods (with a kube2sky in each) causes high load on the docker daemon for each node in the cluster. However this issue still seems to be occurring with a scaled up pod (see below). Is there a misconfiguration here?

```
# sample pod /etc/resolv.conf
nameserver 10.100.100.100
nameserver 10.99.0.2
search default.svc.cluster.local svc.cluster.local cluster.local ec2.internal
options ndots:5
```

```
{
  "apiVersion": "v1",
  "kind": "ReplicationController",
  "metadata": {
    "labels": {
      "k8s-app": "kube-dns",
      "kubernetes.io/cluster-service": "true",
      "version": "v9"
    },
    "name": "kube-dns-v9",
    "namespace": "kube-system"
  },
  "spec": {
    "replicas": 1,
    "selector": {
      "k8s-app": "kube-dns",
      "version": "v9"
    },
    "template": {
      "metadata": {
        "labels": {
          "k8s-app": "kube-dns",
          "kubernetes.io/cluster-service": "true",
          "version": "v9"
        }
      },
      "spec": {
        "containers": [
          {
            "command": [
              "/usr/local/bin/etcd",
              "-data-dir",
              "/var/etcd/data",
              "-listen-client-urls",
              "http://127.0.0.1:2379,http://127.0.0.1:4001",
              "-advertise-client-urls",
              "http://127.0.0.1:2379,http://127.0.0.1:4001",
              "-initial-cluster-token",
              "skydns-etcd"
            ],
            "image": "gcr.io/google_containers/etcd:2.0.9",
            "name": "etcd",
            "resources": {
              "limits": {
                "cpu": "1000m",
                "memory": "500Mi"
              }
            },
            "volumeMounts": [
              {
                "mountPath": "/var/etcd/data",
                "name": "etcd-storage"
              }
            ]
          },
          {
            "args": [
              "-domain=cluster.local"
            ],
            "image": "gcr.io/google_containers/kube2sky:1.11",
            "name": "kube2sky",
            "resources": {
              "limits": {
                "cpu": "200m",
                "memory": "100Mi"
              }
            }
          },
          {
            "args": [
              "-machines=http://localhost:4001",
              "-addr=0.0.0.0:53",
          "-ns-rotate=false",
              "-domain=cluster.local."
            ],
            "image": "gcr.io/google_containers/skydns:2015-10-13-8c72f8c",
            "livenessProbe": {
              "httpGet": {
                "path": "/healthz",
                "port": 8080,
                "scheme": "HTTP"
              },
              "initialDelaySeconds": 30,
              "timeoutSeconds": 5
            },
            "name": "skydns",
            "ports": [
              {
                "containerPort": 53,
                "name": "dns",
                "protocol": "UDP"
              },
              {
                "containerPort": 53,
                "name": "dns-tcp",
                "protocol": "TCP"
              }
            ],
            "readinessProbe": {
              "httpGet": {
                "path": "/healthz",
                "port": 8080,
                "scheme": "HTTP"
              },
              "initialDelaySeconds": 1,
              "timeoutSeconds": 5
            },
            "resources": {
              "limits": {
                "cpu": "2000m",
                "memory": "1000Mi"
              }
            }
          },
          {
            "args": [
              "-cmd=nslookup kubernetes.default.svc.cluster.local. localhost",
              "-port=8080"
            ],
            "image": "gcr.io/google_containers/exechealthz:1.0",
            "name": "healthz",
            "ports": [
              {
                "containerPort": 8080,
                "protocol": "TCP"
              }
            ],
            "resources": {
              "limits": {
                "cpu": "10m",
                "memory": "20Mi"
              }
            }
          }
        ],
        "dnsPolicy": "Default",
        "volumes": [
          {
            "emptyDir": {},
            "name": "etcd-storage"
          }
        ]
      }
    }
  }
}
```

**Comment by AdoHe** [comment]
@fluxrad from the given skydns log, it seems the backend etcd cluster is not reachable. would you please check this?

**Comment by fluxrad** [comment]
Validating that etcd is reachable from the skydns container:

```
$ kubectl exec -t kube-dns-v9-2vszr --namespace=kube-system -c skydns -- wget -O - http://localhost:4001/v2/keys
Connecting to localhost:4001 (127.0.0.1:4001)
{"action":"get","node":{"dir":true,"nodes":[{"key":"/skydns","dir":true,"modifiedIndex":3,"createdIndex":3}]}}
```

Here are the entirety of the logs from the etcd container: 

```
$ kubectl logs kube-dns-v9-2vszr -c etcd --namespace=kube-system
2016/01/14 01:25:13 etcd: listening for peers on http://localhost:2380
2016/01/14 01:25:13 etcd: listening for peers on http://localhost:7001
2016/01/14 01:25:13 etcd: listening for client requests on http://127.0.0.1:2379
2016/01/14 01:25:13 etcd: listening for client requests on http://127.0.0.1:4001
2016/01/14 01:25:13 etcdserver: datadir is valid for the 2.0.1 format
2016/01/14 01:25:13 etcdserver: name = default
2016/01/14 01:25:13 etcdserver: data dir = /var/etcd/data
2016/01/14 01:25:13 etcdserver: member dir = /var/etcd/data/member
2016/01/14 01:25:13 etcdserver: heartbeat = 100ms
2016/01/14 01:25:13 etcdserver: election = 1000ms
2016/01/14 01:25:13 etcdserver: snapshot count = 10000
2016/01/14 01:25:13 etcdserver: advertise client URLs = http://127.0.0.1:2379,http://127.0.0.1:4001
2016/01/14 01:25:13 etcdserver: initial advertise peer URLs = http://localhost:2380,http://localhost:7001
2016/01/14 01:25:13 etcdserver: initial cluster = default=http://localhost:2380,default=http://localhost:7001
2016/01/14 01:25:13 etcdserver: start member 6a5871dbdd12c17c in cluster f68652439e3f8f2a
2016/01/14 01:25:13 raft: 6a5871dbdd12c17c became follower at term 0
2016/01/14 01:25:13 raft: newRaft 6a5871dbdd12c17c [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2016/01/14 01:25:13 raft: 6a5871dbdd12c17c became follower at term 1
2016/01/14 01:25:13 etcdserver: added local member 6a5871dbdd12c17c [http://localhost:2380 http://localhost:7001] to cluster f68652439e3f8f2a
2016/01/14 01:25:14 raft: 6a5871dbdd12c17c is starting a new election at term 1
2016/01/14 01:25:14 raft: 6a5871dbdd12c17c became candidate at term 2
2016/01/14 01:25:14 raft: 6a5871dbdd12c17c received vote from 6a5871dbdd12c17c at term 2
2016/01/14 01:25:14 raft: 6a5871dbdd12c17c became leader at term 2
2016/01/14 01:25:14 raft.node: 6a5871dbdd12c17c elected leader 6a5871dbdd12c17c at term 2
2016/01/14 01:25:14 etcdserver: published {Name:default ClientURLs:[http://127.0.0.1:2379 http://127.0.0.1:4001]} to cluster f68652439e3f8f2a
2016/01/14 02:47:56 etcdserver: start to snapshot (applied: 10001, lastsnap: 0)
2016/01/14 02:47:56 etcdserver: compacted log at index 10001
2016/01/14 02:47:56 etcdserver: saved snapshot at index 10001
```

**Comment by fluxrad** [comment]
Troubleshooting based on some instructions on a somewhat similar issue on kubernetes/kubernetes#14051 

I scaled DNS to 10 pods. This is a tcpdump from a single pod (so presumably 1/10th the traffic):

```
# tcpdump -i any -n udp dst port 53 | grep "packets received" & sleep 10; kill %1
[1] 20586
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on any, link-type LINUX_SLL (Linux cooked), capture size 262144 bytes
38549 packets captured
42767 packets received by filter
4212 packets dropped by kernel
```

We've switched the lion's share of our DNS to skip the `ndots` generates-extra-queries problem by appending a `.` at the end of our most commonly queried hosts. This is a snapshot of our "local" vs "fqdn" profile now:

```
# tcpdump -i any dst port 53 | cut -d' ' -f 8 2>&1 > trace.out & sleep 10; kill %1
# grep -c ".local" trace.out
6822
# wc -l trace.out
32910 trace.out
```

And are still seeing container restarts across pods:

```
$ kubectl get po --namespace=kube-system | grep -E 'dns|RESTARTS'
NAME                                                   READY     STATUS          RESTARTS   AGE
kube-dns-v9-7gxlc                                      4/4       Running         4          2h
kube-dns-v9-a845x                                      4/4       Running         0          2h
kube-dns-v9-ialfy                                      4/4       Running         5          2h
kube-dns-v9-j4y7u                                      4/4       Running         8          2h
kube-dns-v9-kmzb6                                      4/4       Running         3          2h
kube-dns-v9-o5d89                                      4/4       Running         3          2h
kube-dns-v9-q4jyz                                      4/4       Running         5          2h
kube-dns-v9-szc6j                                      4/4       Running         0          2h
kube-dns-v9-tcd9o                                      4/4       Running         0          2h
kube-dns-v9-zg8fx                                      4/4       Running         0          2h
```

It looks like 6,000qps may be too much for each pod-local etcd to handle? I'm not sure how to proceed here, but I can reliably control work throughput of the services in my cluster by scaling DNS pods up/down. In a 40-node cluster, we need at least 25-30 DNS pods at the moment. If we continue to scale `kube2sky` horizontally then I believe we'll continue to put stress on the Docker daemon for each host (is this accurate?). Scaling vertically doesn't seem to ameliorate etcd timeouts.

Hoping @thockin might have some input here?

**Comment by thockin** [comment]
Is this using the userspace proxy or the iptables proxy?

**Comment by fluxrad** [comment]
This is 1.1.3 with iptables proxy enabled:

```
ExecStart=/opt/bin/kubelet \
  --api_servers=${MASTER_URL} \
  --register-node=true \
  --allow-privileged=true \
  --config=/etc/kubernetes/manifests \
  --cluster_dns=${DNS_SERVICE_IP} \
  --cluster_domain=cluster.local \
  --kubeconfig=/etc/kubernetes/worker-kubeconfig.yaml \
  --tls-cert-file=/etc/kubernetes/ssl/kubernetes-worker.crt \
  --tls-private-key-file=/etc/kubernetes/ssl/kubernetes-worker.key \
  --cloud-provider=aws \
  --cadvisor-port=0
```

**Comment by thockin** [comment]
Can you run `iptables-save` on a node for me?

**Comment by fluxrad** [comment]
See below:


```
# Generated by iptables-save v1.4.21 on Tue Jan 19 22:37:30 2016
*nat
:PREROUTING ACCEPT [2959:369092]
:INPUT ACCEPT [1635:267255]
:OUTPUT ACCEPT [1643:270557]
:POSTROUTING ACCEPT [1640:270327]
:DOCKER - [0:0]
:FLANNEL - [0:0]
:KUBE-NODEPORTS - [0:0]
:KUBE-SEP-5DKUDHJNXYVADW4W - [0:0]
:KUBE-SEP-7PLZTBUJCZOWT5SO - [0:0]
:KUBE-SEP-AJ4Q3PT3REHTEYA7 - [0:0]
:KUBE-SEP-BPLNEZ4YWBOKUQQ2 - [0:0]
:KUBE-SEP-DB2A5KY72PCDR2MR - [0:0]
:KUBE-SEP-DVOOBOWGUT4AFI7Y - [0:0]
:KUBE-SEP-GQY3WTNOEYM2STTC - [0:0]
:KUBE-SEP-I7GXSU6M5S576YEF - [0:0]
:KUBE-SEP-RRO4YIQYURK3W3RX - [0:0]
:KUBE-SEP-T2SF67SAYLY2LHEU - [0:0]
:KUBE-SEP-TOGN32YQROSDHOQO - [0:0]
:KUBE-SEP-TVPHEOWIXACEDWXZ - [0:0]
:KUBE-SEP-UKKWJFFDJ43ERL2R - [0:0]
:KUBE-SEP-USB6NTUHGB5UJFRF - [0:0]
:KUBE-SERVICES - [0:0]
:KUBE-SVC-6N4SJQIF3IX3FORG - [0:0]
:KUBE-SVC-BJM46V3U5RZHCFRZ - [0:0]
:KUBE-SVC-CK6HVV5A27TDFNIA - [0:0]
:KUBE-SVC-ERIFXISQEP7F7OF4 - [0:0]
:KUBE-SVC-JRXTEHDDTAFMSEAS - [0:0]
:KUBE-SVC-OPFQXZDFMZWLGLGG - [0:0]
:KUBE-SVC-TCOU7JCQXEZGVUNU - [0:0]
:KUBE-SVC-WEHLQ23XZWSA5ZX3 - [0:0]
-A PREROUTING -m comment --comment "kubernetes service portals" -j KUBE-SERVICES
-A PREROUTING -m addrtype --dst-type LOCAL -j DOCKER
-A OUTPUT -m comment --comment "kubernetes service portals" -j KUBE-SERVICES
-A OUTPUT ! -d 127.0.0.0/8 -m addrtype --dst-type LOCAL -j DOCKER
-A POSTROUTING -s 10.10.0.0/16 -j FLANNEL
-A POSTROUTING -m comment --comment "kubernetes service traffic requiring SNAT" -m mark --mark 0x4d415351 -j MASQUERADE
-A FLANNEL -d 10.10.0.0/16 -j ACCEPT
-A FLANNEL ! -d 224.0.0.0/4 -j MASQUERADE
-A KUBE-NODEPORTS -p tcp -m comment --comment "kube-system/monitoring-grafana:" -m tcp --dport 30001 -j MARK --set-xmark 0x4d415351/0xffffffff
-A KUBE-NODEPORTS -p tcp -m comment --comment "kube-system/monitoring-grafana:" -m tcp --dport 30001 -j KUBE-SVC-JRXTEHDDTAFMSEAS
-A KUBE-SEP-5DKUDHJNXYVADW4W -s 10.10.50.77/32 -m comment --comment "kube-system/kube-dns:dns-tcp" -j MARK --set-xmark 0x4d415351/0xffffffff
-A KUBE-SEP-5DKUDHJNXYVADW4W -p tcp -m comment --comment "kube-system/kube-dns:dns-tcp" -m tcp -j DNAT --to-destination 10.10.50.77:53
-A KUBE-SEP-7PLZTBUJCZOWT5SO -s 10.10.142.56/32 -m comment --comment "kube-system/kube-dns:dns" -j MARK --set-xmark 0x4d415351/0xffffffff
-A KUBE-SEP-7PLZTBUJCZOWT5SO -p udp -m comment --comment "kube-system/kube-dns:dns" -m udp -j DNAT --to-destination 10.10.142.56:53
-A KUBE-SEP-AJ4Q3PT3REHTEYA7 -s 10.10.95.51/32 -m comment --comment "kube-system/kubedash:" -j MARK --set-xmark 0x4d415351/0xffffffff
-A KUBE-SEP-AJ4Q3PT3REHTEYA7 -p tcp -m comment --comment "kube-system/kubedash:" -m tcp -j DNAT --to-destination 10.10.95.51:8289
-A KUBE-SEP-BPLNEZ4YWBOKUQQ2 -s 10.10.50.77/32 -m comment --comment "kube-system/kube-dns:dns" -j MARK --set-xmark 0x4d415351/0xffffffff
-A KUBE-SEP-BPLNEZ4YWBOKUQQ2 -p udp -m comment --comment "kube-system/kube-dns:dns" -m udp -j DNAT --to-destination 10.10.50.77:53
-A KUBE-SEP-DB2A5KY72PCDR2MR -s 10.10.42.84/32 -m comment --comment "kube-system/monitoring-grafana:" -j MARK --set-xmark 0x4d415351/0xffffffff
-A KUBE-SEP-DB2A5KY72PCDR2MR -p tcp -m comment --comment "kube-system/monitoring-grafana:" -m tcp -j DNAT --to-destination 10.10.42.84:3000
-A KUBE-SEP-DVOOBOWGUT4AFI7Y -s 10.10.119.172/32 -m comment --comment "kube-system/kube-dns:dns" -j MARK --set-xmark 0x4d415351/0xffffffff
-A KUBE-SEP-DVOOBOWGUT4AFI7Y -p udp -m comment --comment "kube-system/kube-dns:dns" -m udp -j DNAT --to-destination 10.10.119.172:53
-A KUBE-SEP-GQY3WTNOEYM2STTC -s 10.10.88.52/32 -m comment --comment "kube-system/kube-dns:dns-tcp" -j MARK --set-xmark 0x4d415351/0xffffffff
-A KUBE-SEP-GQY3WTNOEYM2STTC -p tcp -m comment --comment "kube-system/kube-dns:dns-tcp" -m tcp -j DNAT --to-destination 10.10.88.52:53
-A KUBE-SEP-I7GXSU6M5S576YEF -s 10.10.56.97/32 -m comment --comment "kube-system/kube-dns:dns" -j MARK --set-xmark 0x4d415351/0xffffffff
-A KUBE-SEP-I7GXSU6M5S576YEF -p udp -m comment --comment "kube-system/kube-dns:dns" -m udp -j DNAT --to-destination 10.10.56.97:53
-A KUBE-SEP-RRO4YIQYURK3W3RX -s 10.10.88.52/32 -m comment --comment "kube-system/kube-dns:dns" -j MARK --set-xmark 0x4d415351/0xffffffff
-A KUBE-SEP-RRO4YIQYURK3W3RX -p udp -m comment --comment "kube-system/kube-dns:dns" -m udp -j DNAT --to-destination 10.10.88.52:53
-A KUBE-SEP-T2SF67SAYLY2LHEU -s 10.10.95.50/32 -m comment --comment "kube-system/heapster:" -j MARK --set-xmark 0x4d415351/0xffffffff
-A KUBE-SEP-T2SF67SAYLY2LHEU -p tcp -m comment --comment "kube-system/heapster:" -m tcp -j DNAT --to-destination 10.10.95.50:8082
-A KUBE-SEP-TOGN32YQROSDHOQO -s 10.10.56.97/32 -m comment --comment "kube-system/kube-dns:dns-tcp" -j MARK --set-xmark 0x4d415351/0xffffffff
-A KUBE-SEP-TOGN32YQROSDHOQO -p tcp -m comment --comment "kube-system/kube-dns:dns-tcp" -m tcp -j DNAT --to-destination 10.10.56.97:53
-A KUBE-SEP-TVPHEOWIXACEDWXZ -s 10.50.27.58/32 -m comment --comment "default/kubernetes:" -j MARK --set-xmark 0x4d415351/0xffffffff
-A KUBE-SEP-TVPHEOWIXACEDWXZ -p tcp -m comment --comment "default/kubernetes:" -m tcp -j DNAT --to-destination 10.50.27.58:443
-A KUBE-SEP-UKKWJFFDJ43ERL2R -s 10.10.119.172/32 -m comment --comment "kube-system/kube-dns:dns-tcp" -j MARK --set-xmark 0x4d415351/0xffffffff
-A KUBE-SEP-UKKWJFFDJ43ERL2R -p tcp -m comment --comment "kube-system/kube-dns:dns-tcp" -m tcp -j DNAT --to-destination 10.10.119.172:53
-A KUBE-SEP-USB6NTUHGB5UJFRF -s 10.10.142.56/32 -m comment --comment "kube-system/kube-dns:dns-tcp" -j MARK --set-xmark 0x4d415351/0xffffffff
-A KUBE-SEP-USB6NTUHGB5UJFRF -p tcp -m comment --comment "kube-system/kube-dns:dns-tcp" -m tcp -j DNAT --to-destination 10.10.142.56:53
-A KUBE-SERVICES -d 10.99.0.1/32 -p tcp -m comment --comment "default/kubernetes: cluster IP" -m tcp --dport 443 -j KUBE-SVC-6N4SJQIF3IX3FORG
-A KUBE-SERVICES -d 10.99.27.10/32 -p tcp -m comment --comment "kube-system/heapster: cluster IP" -m tcp --dport 80 -j KUBE-SVC-BJM46V3U5RZHCFRZ
-A KUBE-SERVICES -d 10.99.100.100/32 -p udp -m comment --comment "kube-system/kube-dns:dns cluster IP" -m udp --dport 53 -j KUBE-SVC-TCOU7JCQXEZGVUNU
-A KUBE-SERVICES -d 10.99.100.100/32 -p tcp -m comment --comment "kube-system/kube-dns:dns-tcp cluster IP" -m tcp --dport 53 -j KUBE-SVC-ERIFXISQEP7F7OF4
-A KUBE-SERVICES -d 10.99.178.153/32 -p tcp -m comment --comment "kube-system/kubedash: cluster IP" -m tcp --dport 80 -j KUBE-SVC-OPFQXZDFMZWLGLGG
-A KUBE-SERVICES -d 10.99.53.59/32 -p tcp -m comment --comment "kube-system/monitoring-influxdb:http cluster IP" -m tcp --dport 8083 -j KUBE-SVC-WEHLQ23XZWSA5ZX3
-A KUBE-SERVICES -d 10.99.53.59/32 -p tcp -m comment --comment "kube-system/monitoring-influxdb:api cluster IP" -m tcp --dport 8086 -j KUBE-SVC-CK6HVV5A27TDFNIA
-A KUBE-SERVICES -d 10.99.129.121/32 -p tcp -m comment --comment "kube-system/monitoring-grafana: cluster IP" -m tcp --dport 80 -j KUBE-SVC-JRXTEHDDTAFMSEAS
-A KUBE-SERVICES -m comment --comment "kubernetes service nodeports; NOTE: this must be the last rule in this chain" -m addrtype --dst-type LOCAL -j KUBE-NODEPORTS
-A KUBE-SVC-6N4SJQIF3IX3FORG -m comment --comment "default/kubernetes:" -j KUBE-SEP-TVPHEOWIXACEDWXZ
-A KUBE-SVC-BJM46V3U5RZHCFRZ -m comment --comment "kube-system/heapster:" -j KUBE-SEP-T2SF67SAYLY2LHEU
-A KUBE-SVC-ERIFXISQEP7F7OF4 -m comment --comment "kube-system/kube-dns:dns-tcp" -m statistic --mode random --probability 0.20000000019 -j KUBE-SEP-UKKWJFFDJ43ERL2R
-A KUBE-SVC-ERIFXISQEP7F7OF4 -m comment --comment "kube-system/kube-dns:dns-tcp" -m statistic --mode random --probability 0.25000000000 -j KUBE-SEP-USB6NTUHGB5UJFRF
-A KUBE-SVC-ERIFXISQEP7F7OF4 -m comment --comment "kube-system/kube-dns:dns-tcp" -m statistic --mode random --probability 0.33332999982 -j KUBE-SEP-5DKUDHJNXYVADW4W
-A KUBE-SVC-ERIFXISQEP7F7OF4 -m comment --comment "kube-system/kube-dns:dns-tcp" -m statistic --mode random --probability 0.50000000000 -j KUBE-SEP-TOGN32YQROSDHOQO
-A KUBE-SVC-ERIFXISQEP7F7OF4 -m comment --comment "kube-system/kube-dns:dns-tcp" -j KUBE-SEP-GQY3WTNOEYM2STTC
-A KUBE-SVC-JRXTEHDDTAFMSEAS -m comment --comment "kube-system/monitoring-grafana:" -j KUBE-SEP-DB2A5KY72PCDR2MR
-A KUBE-SVC-OPFQXZDFMZWLGLGG -m comment --comment "kube-system/kubedash:" -j KUBE-SEP-AJ4Q3PT3REHTEYA7
-A KUBE-SVC-TCOU7JCQXEZGVUNU -m comment --comment "kube-system/kube-dns:dns" -m statistic --mode random --probability 0.20000000019 -j KUBE-SEP-DVOOBOWGUT4AFI7Y
-A KUBE-SVC-TCOU7JCQXEZGVUNU -m comment --comment "kube-system/kube-dns:dns" -m statistic --mode random --probability 0.25000000000 -j KUBE-SEP-7PLZTBUJCZOWT5SO
-A KUBE-SVC-TCOU7JCQXEZGVUNU -m comment --comment "kube-system/kube-dns:dns" -m statistic --mode random --probability 0.33332999982 -j KUBE-SEP-BPLNEZ4YWBOKUQQ2
-A KUBE-SVC-TCOU7JCQXEZGVUNU -m comment --comment "kube-system/kube-dns:dns" -m statistic --mode random --probability 0.50000000000 -j KUBE-SEP-I7GXSU6M5S576YEF
-A KUBE-SVC-TCOU7JCQXEZGVUNU -m comment --comment "kube-system/kube-dns:dns" -j KUBE-SEP-RRO4YIQYURK3W3RX
COMMIT
# Completed on Tue Jan 19 22:37:30 2016
# Generated by iptables-save v1.4.21 on Tue Jan 19 22:37:30 2016
*filter
:INPUT ACCEPT [338425257:58826840575]
:FORWARD ACCEPT [95493214:8523126528]
:OUTPUT ACCEPT [338408249:87382745697]
:DOCKER - [0:0]
-A FORWARD -o docker0 -j DOCKER
-A FORWARD -o docker0 -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A FORWARD -i docker0 ! -o docker0 -j ACCEPT
-A FORWARD -i docker0 -o docker0 -j ACCEPT
COMMIT
#
```

**Comment by ArtfulCoder** [comment]
can you try with kube-dns-v10. (latest version of kubernetes should give you that) ?

you are using an older version, where the resolv.conf has multiple nameservers instead of just the kubernetes dns server.

**Comment by ArtfulCoder** [comment]
We also picked a newer skydns server that has a bug fix in kube-dns-v10..

**Comment by fluxrad** [comment]
Yeah, using v10 was the first thing I tried, with the same results: frequent io timouts talking to etcd. I moved back to the container configuration specified in the `release-1.1` branch since I assumed this was the supported latest. I'll move us back to kube-dns-v10.

**Comment by ArtfulCoder** [comment]
when you moved to kube-dns-v10, can you confirm that the pod had only one nameserver in the resolv.conf of the pod containers ?

**Comment by ArtfulCoder** [comment]
also, to summarize your observation, for 1000 pod cluster, 40 dns pods work reliably ?

**Comment by fluxrad** [comment]
I validated that `/etc/resolv.conf` doesn't appear to be getting updated on new RC's/pods after bumping DNS to v10. Do I have to bounce my kubelets?

```
kubectl exec -t busybox-7ktjg  -- cat /etc/resolv.conf
nameserver 10.100.100.100
nameserver 10.99.0.2
search default.svc.cluster.local svc.cluster.local cluster.local ec2.internal
options ndots:5
```

This is the current setup:

```
{
  "apiVersion": "v1",
  "kind": "ReplicationController",
  "metadata": {
    "labels": {
      "k8s-app": "kube-dns",
      "kubernetes.io/cluster-service": "true",
      "version": "v10"
    },
    "name": "kube-dns-v10",
    "namespace": "kube-system"
  },
  "spec": {
    "replicas": 1,
    "selector": {
      "k8s-app": "kube-dns",
      "version": "v10"
    },
    "template": {
      "metadata": {
        "labels": {
          "k8s-app": "kube-dns",
          "kubernetes.io/cluster-service": "true",
          "version": "v10"
        }
      },
      "spec": {
        "containers": [
          {
            "command": [
              "/usr/local/bin/etcd",
              "-data-dir",
              "/var/etcd/data",
              "-listen-client-urls",
              "http://127.0.0.1:2379,http://127.0.0.1:4001",
              "-advertise-client-urls",
              "http://127.0.0.1:2379,http://127.0.0.1:4001",
              "-initial-cluster-token",
              "skydns-etcd"
            ],
            "image": "gcr.io/google_containers/etcd:2.0.9",
            "name": "etcd",
            "resources": {
              "limits": {
                "cpu": "500m",
                "memory": "256Mi"
              }
            },
            "volumeMounts": [
              {
                "mountPath": "/var/etcd/data",
                "name": "etcd-storage"
              }
            ]
          },
          {
            "args": [
              "-domain=cluster.local"
            ],
            "image": "gcr.io/google_containers/kube2sky:1.12",
            "name": "kube2sky",
            "resources": {
              "limits": {
                "cpu": "200m",
                "memory": "100Mi"
              }
            }
          },
          {
            "args": [
              "-machines=http://localhost:4001",
              "-addr=0.0.0.0:53",
	      "-ns-rotate=false",
              "-domain=cluster.local."
            ],
            "image": "gcr.io/google_containers/skydns:2015-10-13-8c72f8c",
            "livenessProbe": {
              "httpGet": {
                "path": "/healthz",
                "port": 8080,
                "scheme": "HTTP"
              },
              "initialDelaySeconds": 30,
              "timeoutSeconds": 5
            },
            "name": "skydns",
            "ports": [
              {
                "containerPort": 53,
                "name": "dns",
                "protocol": "UDP"
              },
              {
                "containerPort": 53,
                "name": "dns-tcp",
                "protocol": "TCP"
              }
            ],
            "readinessProbe": {
              "httpGet": {
                "path": "/healthz",
                "port": 8080,
                "scheme": "HTTP"
              },
              "initialDelaySeconds": 1,
              "timeoutSeconds": 5
            },
            "resources": {
              "limits": {
                "cpu": "1000m",
                "memory": "512Mi"
              }
            }
          },
          {
            "args": [
              "-cmd=nslookup kubernetes.default.svc.cluster.local localhost",
              "-port=8080"
            ],
            "image": "gcr.io/google_containers/exechealthz:1.0",
            "name": "healthz",
            "ports": [
              {
                "containerPort": 8080,
                "protocol": "TCP"
              }
            ],
            "resources": {
              "limits": {
                "cpu": "10m",
                "memory": "20Mi"
              }
            }
          }
        ],
        "dnsPolicy": "Default",
        "volumes": [
          {
            "emptyDir": {},
            "name": "etcd-storage"
          }
        ]
      }
    }
  }
}
```

Also, your summary is accurate. When I scale the number of DNS pods to ~40 (or, roughly 1:1 nodes-to-dns) I see acceptable DNS performance, with few — if any — dns pods showing restarts.

**Comment by ArtfulCoder** [comment]
I think you have a version where kubelet is not generating the newer resolv.conf which has only one ns.

I believe the one ns per pod in resolv.conf was part of 1.1 

I am surprised that your pods have more than 1 ns.

**Comment by ArtfulCoder** [comment]
maybe, you could create a new pod and check how many ns you have in the resolv.conf

**Comment by fluxrad** [comment]
Yup, same issue w/ kubelet 1.1.3, dns-v10 and a new RC. 

```
$ kubectl delete rc busybox
$ kubectl run busybox --image=busybox sleep 3600
$ kubectl get po | grep busybox
busybox-7ktjg                   1/1       Terminating   0          49m
busybox-gq3n5                   1/1       Running       0          14s
$ kubectl exec -t busybox-gq3n5 -- cat /etc/resolv.conf
nameserver 10.100.100.100
nameserver 10.99.0.2
search default.svc.cluster.local svc.cluster.local cluster.local ec2.internal
options ndots:5
```

And validating the kubelet version

```
$ kubelet --version=true
Kubernetes v1.1.3
```

NB I haven't modified my node DNS at all. Also, for the moment I've switched all of my production pods to use `dnsPolicy: Default` in order to avoid the DNS bottleneck. 

```
random-node$ cat /etc/resolv.conf
nameserver 10.99.0.2
search ec2.internal
```

**Comment by ArtfulCoder** [comment]
yes,  I verified that the nameserver change is not yet in.
"git branch -a --contains bb3f49" shows that its only in master..

**Comment by fluxrad** [comment]
Is there a target release for that code? I'm unclear how removing the secondary DNS server from `/etc/resolv.conf` across my containers will fix load/connectivity issues on the DNS pods themselves.

**Comment by ArtfulCoder** [comment]
I was just trying to understand your setup, so that I understood what issue you could be facing.
Updating the etc/resolv.con won't help with load issues.

**Comment by ArtfulCoder** [comment]
I am coming up with a new version of kubedns pod and will have local dns caching on nodes. I will keep this issue open till then

**Comment by rakesh-k** [comment]
FWIW I am running kube-dns-v11 and seeing the same error cropping up in the skydns log:



```
2016-05-07T04:23:51.203991261Z 2016/05/07 04:23:51 skydns: ready for queries on toto. for udp://0.0.0.0:53 [rcache 0]
2016-05-07T04:23:51.440137614Z 2016/05/07 04:23:51 skydns: error from backend: 501: All the given peers are not reachable (failed to propose on members [http://127.0.0.1:4001] twice [last error: Get http://127.0.0.1:4001/v2/keys/skydns/toto/svc/elasticsearch-logging?quorum=false&recursive=true&sorted=false: dial tcp 127.0.0.1:4001: connection refused]) [0]
```

**Comment by fluxrad** [comment]
It looks like the caching issue has helped, but we're still seeing a huge DNS load due to the `ndots:5` setting in the cluster. In many cases, we now see `nf_conntrack: table full, dropping packet` errors, wiht some nodes tracking over 300k DNS connections through iptables:

### Node and process counts:

```
# AWS m4.2xlarge 8 core @ 32GB memory
$ kubectl get no | wc -l
      71
```

```
$ kubectl get rc  --namespace=kube-system -l k8s-app=kube-dns
NAME           DESIRED   CURRENT   AGE
kube-dns-v18   10        10        48d
```

When I run the conntrack tool against machines running kube-dns pods, I notice an excessive number of DNS connections tracked:

```
# conntrack -C
519928

# conntrack -L -p udp --dport 53 | wc -l
conntrack v1.4.2 (conntrack-tools): 305746 flow entries have been shown.
305746
```

We've gone so far as setting `--conntrack-max-per-core=65536` inside kube-proxy. Not sure how to proceed here, we're no longer memory/cpu constrained, but are now limited by the number of connections iptables can track within the cluster.

**Comment by thockin** [comment]
@girishkalele  @mrhohn

**Comment by girishkalele** [comment]
@fluxrad 

I had an experimental PR #28913 out to inject a `use-vc` setting into the resolv.conf to force TCP.

With UDP, there is no way to "inform" conntrack that the session/connection is over, but with TCP, the FIN/RST would help conntrack free up slots as soon as possible.

Unfortunately, we have not had the bandwidth to test it out. It also needs some code factoring and maybe a feature gate flag before we can commit it.

#28913

**Comment by girishkalele** [comment]
Looking at dnsmasq caching behaviour, it does not cache NXDOMAIN results for any length of time, so the same repeated ndots searches keep going through and hitting kube-dns all the time.

One possible enhancement (to dnsmasq) is to allow NXDOMAIN caching with a programmable TTL.

**Comment by fluxrad** [comment]
Caching NXDOMAIN's would probably be a step in the right direction here. I imagine creating and tearing down connections for each NXDOMAIN request would get quite expensive. 

Alternatively, is there any potential support for making `ndots` tunable via the kubelet? I'd be comfortable enforcing stricter internal lookup requirements (e.g. all internal requests must use strict FQDN, or at least the first three parts: `foo.default.svc`) if it meant eliminating 3/4 of my DNS traffic.

**Comment by thockin** [comment]
Changing search paths would be a VERY large scale change and would be
subject to standard deprecation policy.  Even so, changing it as you
suggest would require ndots:3, which would STILL match the vast majority of
lookups.  "google.com", "mail.google.com", "finance.yahoo.com", "
www.mcdonalds.com", "storage.googleapis.com", etc.  The real win would be
to get to ndots:1 or ndots:2.  See previous remarks about deprecation.

WRT caching NXDOMAIN - for how long?  It has to be relatively short.

**Comment by fluxrad** [comment]
Re `ndots` tuning, I was under the impression that this wouldn't impact core Kubernetes functionality since DNS is an add-on — this may be a false impression. Assuming a switch like `--container-ndots` with a (sensible) default of `5`, would this provide the normal functionality but give the end-user the ability to override in high-query-volume use-cases? 

Really all we're trying to solve for is a solution that prevents generating large amounts of unnecessary DNS traffic, which can quickly swamp individual nodes' connection tracking capabilities. At 70 nodes with 30 DNS pods, we still occasionally see nodes saturating a 512k nf_conntrack_max (double the default).  One current solution has been to use `dnsPolicy: Default` and do service discovery through ENVs. Ideally, of course, everything would be using cluster DNS first.

Re: caching NXDOMAINs. I would assume somewhere around a 10-60s timeout for caching NXDOMAIN would be reasonable?

**Comment by thockin** [comment]
On Wed, Sep 7, 2016 at 4:15 PM, Aaron Mills <notifications@github.com> wrote:
>
> Re ndots tuning, I was under the impression that this wouldn't impact core Kubernetes functionality since DNS is an add-on — this may be a false impression. Assuming a switch like --container-ndots with a (sensible) default of 5, would this provide the normal functionality but give the end-user the ability to override in high-query-volume use-cases?

It was an addon.  DNS is effectively required at this point.  You
could MAYBE make an argument that ndots is a cluster config parameter,
but doing so negates a lot of portability, so I really don't think it
is smart to change from the default.  If there was enough pressure I
could maybe be convinced to flag it, but I would advise against anyone
changing it.

> Really all we're trying to solve for is a solution that prevents generating large amounts of unnecessary DNS traffic, which can quickly swamp individual nodes' connection tracking capabilities. At 70 nodes with 30 DNS pods, we still occasionally see nodes saturating a 512k nf_conntrack_max (double the default). One current solution has been to use dnsPolicy: Default and do service discovery through ENVs. Ideally, of course, everything would be using cluster DNS first.

We have some amount of parameterization we can control.  We can offer
more conntrack entries.  We can shorten the UDP timeout.  Maybe we can
use some node-local go-between that doesn't trigger conntrack and does
socket-reuse or even doesn't trigger conntrack itself.

> Re: caching NXDOMAINs. I would assume somewhere around a 10-60s timeout for caching NXDOMAIN would be reasonable?

60s feels on the high-end, but not out of bounds.  I think we would
have to do some literature exploration.  We're not the first to get
here :)

**Comment by macb** [comment]
Stumbled across this. I think `ndots` really should be a configurable:
https://github.com/kubernetes/kubernetes/issues/33554#issuecomment-251755257

**Comment by macb** [comment]
I think `skydns` is also doing a very poor job of caching on its end. Noticed I was getting some very odd metrics (see https://github.com/skynetservices/skydns/issues/298) and I think it may be sending from [here](https://github.com/skynetservices/skydns/blob/master/server/server.go#L283-L285) and so never actually inserts into its own cache [here](https://github.com/skynetservices/skydns/blob/master/server/server.go#L287). Not 100% sure without some better logging or metrics to track it all down but seemed like it may be part of poor performance.

**Comment by thockin** [comment]
For kubernetes v1.3 and v.14 we have added a dnsmasq container in the
chain, as a better cache.

**Comment by macb** [comment]
Ah, I do recall seeing that. Time to find out how to see how effective it is!

**Comment by fluxrad** [comment]
FWIW turned on prometheus metrics in skydns and found the following. I'm a little confused as to why error rates are significantly higher than overall request counts:

```
# per-second values for a single pod

rate(skydns_skydns_dns_request_count_total{kubernetes_name="kube-dns-v19-cffk0"}[5m]) 

{ system: recursive} - 39.366 / second
{ system: reverse} - 0.53 / second

rate(skydns_skydns_dns_error_count_total{kubernetes_name="kube-dns-v19-cffk0"}[5m])

{ cause: nxdomain, system: auth } - 230.04 / second
{ cause: nxdomain, system: recursive} - 37.87 / second
{ cause: servfail, system: recursive} - 0.64 / second

rate(skydns_skydns_dns_cachemiss_count_total{kubernetes_name="kube-dns-v19-cffk0"}[5m])

{ cache: response } - 95.61 / second
```

### Query data

```
# total qeries /second sample
# sysdig -w saved.cap -M 30
# sysdig -r saved.cap "proc.name=dnsmasq and evt.category=net and fd.port=10053 and evt.dir=>" | grep -c "recvfrom"

Results:

Sample node 1: 136 qps
Sample node 2: 506 qps
```

While the switch to dnsmasq has ameliorated some issues, it does look like this may be causing excessive load on AWS itself. We're now seeing tens of DNS timeouts/second talking to AWS DNS from every DNS pod:

```
# ips changed
kubectl logs kube-dns-v19-cffk0 --namespace=kube-system --tail=10  -c skydns
I1012 23:19:53.300427       1 logs.go:41] skydns: failure to forward request "read udp 10.10.10.11:40790->10.9.9.9:53: i/o timeout"
I1012 23:19:53.300499       1 logs.go:41] skydns: failure to forward request "read udp 10.10.10.11:40790->10.9.9.9:53: i/o timeout"
I1012 23:19:53.300527       1 logs.go:41] skydns: failure to forward request "read udp 10.10.10.11:40790->10.9.9.9:53: i/o timeout"
I1012 23:19:53.300554       1 logs.go:41] skydns: failure to forward request "read udp 10.10.10.11:40790->10.9.9.9:53: i/o timeout"
I1012 23:19:53.300580       1 logs.go:41] skydns: failure to forward request "read udp 10.10.10.11:40790->10.9.9.9:53: i/o timeout"
I1012 23:19:54.809638       1 logs.go:41] skydns: failure to forward request "read udp 10.10.10.11:60697->10.9.9.9:53: i/o timeout"
I1012 23:19:58.133050       1 logs.go:41] skydns: failure to forward request "read udp 10.10.10.11:52286->10.9.9.9:53: i/o timeout"
I1012 23:20:00.258232       1 logs.go:41] skydns: failure to forward request "read udp 10.10.10.11:36838->10.9.9.9:53: i/o timeout"
I1012 23:20:03.794403       1 logs.go:41] skydns: failure to forward request "read udp 10.10.10.11:37153->10.9.9.9:53: i/o timeout"
I1012 23:20:04.729122       1 logs.go:41] skydns: failure to forward request "read udp 10.10.10.11:55770->10.9.9.9:53: i/o timeout"
```

I've also opened a ticket with AWS to understand their DNS limits. Am I correct in assuming that queries like `www.google.com.default.svc.cluster.local` should cause skydns to return an NXDOMAIN but not actually do a lookup outside the cluster?

**Comment by macb** [comment]
@fluxrad I think I ran into the same quirky metrics and submitted https://github.com/skynetservices/skydns/pull/297 which I hope at least brings them somewhat closer but haven't actually looked into upgrading it in `kube-dns`.

I think it may also report double the errors which I filed an issue for https://github.com/skynetservices/skydns/issues/298.

**Comment by fluxrad** [comment]
More data here. I ran a sysdig against one system hosting a DNS pod. And found an order of magnitude more bogus queries being generated. I'm assuming this is because skydns isn't caching NXDOMAIN records, but is caching valid entries:

```
# 30 second capture
root@6c13fd7476d6# sysdig -w output.scap -M 30

# NB: Almost all queries ending in ec2.internal are bogus (search ec2.local in /etc/resolv.conf)
# He we're sending approximately 40 bad queries/second to AWS.
root@6c13fd7476d6:/# sysdig -r output.scap fd.ip=10.9.9.9 and fd.port=53 and proc.name=kube-dns | grep "< read" | grep -v EAGAIN | grep -c "ec2.internal"
1228

# Total valid queries sent to AWS: ~ 5/second
root@6c13fd7476d6:/# sysdig -r output.scap fd.ip=10.9.9.9 and fd.port=53 and proc.name=kube-dns | grep "< read" | grep -v EAGAIN | grep -c -v "ec2.internal"
149

# Total queries sent through from dnsmasq to skydns: ~ 186/second
# sysdig -r output.scap fd.ip=127.0.0.1 and fd.port=10053 and evt.type=sendto | wc -l
5567

# Total queries received by the pod
# sysdig -r output.scap "proc.name=dnsmasq and fd.cport=53 and evt.type=recvmsg and evt.dir=>" | wc -l
6713
```

Someone should double check my methods, but the above suggests that 80% of queries are falling through dnsmasq, and of those 22% are propagating bogus queries to AWS. A separate 61% appear to be bogus queries to Kubernetes, e.g. mydomain.com.cluster.local, etc.:

```
# queries to "*.cluster.local" that propagate through to skydns (NXDOMAIN)
root@6c13fd7476d6:/# sysdig -r output.scap fd.ip=127.0.0.1 and fd.port=10053 and evt.type=sendto | grep -c "cluster.local"
4104
```

I did check to see if dnsmasq had a "cache NXDOMAIN" setting by default, but I didn't see it. I do see that you can return records for specific domains/subdomains. We may try this for domains like `mydomain.com.cluster.local` and see how it effects overall DNS performance:

```
-A, --address=/<domain>/[domain/][<ipaddr>]
Specify an IP address to return for any host in the given domains. Queries in the domains are never forwarded and always replied to with the specified IP address which may be IPv4 or IPv6. To give both IPv4 and IPv6 addresses for a domain, use repeated -A flags. Note that /etc/hosts and DHCP leases override this for individual names. A common use of this is to redirect the entire doubleclick.net domain to some friendly local web server to avoid banner ads. The domain specification works in the same was as for --server, with the additional facility that /#/ matches any domain. Thus --address=/#/1.2.3.4 will always return 1.2.3.4 for any query not answered from /etc/hosts or DHCP and not sent to an upstream nameserver by a more specific --server directive. As for --server, one or more domains with no address returns a no-such-domain answer, so --address=/example.com/ is equivalent to --server=/example.com/ and returns NXDOMAIN for example.com and all its subdomains.
```

**Comment by fluxrad** [comment]
Looks like using the `-A` flag with dnsmasq relieves quite a bit of pressure on Skydns:

#### dnsmaq args:
```
        - --cache-size=64000
        - --no-resolv
        - --server=127.0.0.1#10053
        # respond with NXDOMAIN for the following domains (and their subdomains)
        - --address=/com.cluster.local/com.svc.cluster.local/com.default.svc.cluster.local/<ourdomainname>.cluster.local/<ourdomainname>.svc.cluster.local/<ourdomainname>.default.svc.cluster.local/<ourdomainname>.ec2.internal/com.ec2.internal/
```

#### New test

```
# Total queries sent
root@ca8ab049ed5f:/# sysdig -r output.scap "proc.name=dnsmasq and fd.sport=53 and evt.type=recvmsg and evt.dir=>" | wc -l
4575

# Total queries sent through to skydns: 7%
root@ca8ab049ed5f:/# sysdig -r output.scap fd.ip=127.0.0.1 and fd.port=10053 and evt.type=sendto | wc -l
322

# Total good queries sent through to AWS: approx. 2%
root@ca8ab049ed5f:/# sysdig -r output.scap fd.ip=10.9.9.9 and fd.port=53 and proc.name=kube-dns | grep "< read" | grep -v EAGAIN | grep -c -v "ec2.internal"
94

# Total bogus queries sent through to AWS
root@ca8ab049ed5f:/# sysdig -r output.scap fd.ip=10.9.9.9 and fd.port=53 and proc.name=kube-dns | grep "< read" | grep -v EAGAIN | grep -c -E "ec2.internal|amazonaws.com"
54
```

I'm still seeing timeouts talking to AWS so there may be a separate issue there. We still probably have a number of bogus queries that slip through and don't resolve, e.g. `google.com.au.default.svc.cluster.local` or what have you. Ameliorating DNS load in this way still feels clunky.

**Comment by paultyng** [comment]
This affected our cluster pretty hard recently with new relic.  New relic would try every search domain for every request so hammered dns.  Once we added the `--address` setting to return `NXDOMAIN`, the logs slowed down substantially.  We haven't tried this in prod yet, but just observed anecdotally.

**Comment by vladlosev** [comment]
I recently started seeing similar problems although with a smaller cluster. Our code uses a lot of host names without suffixes, which results in a lot of `NXDOMAIN`-returning lookups. After enabling logging in dnsmasq we have found that it only caches `NXDOMAIN` results from the upstream provider, never from kube-dns. E.g.:
```
Vlads-MacBook-Air:~ vlad$ kc --namespace=kube-system logs kube-dns-885961120-kxhco dnsmasq -f --tail=10 | grep NXDOMAIN | head -n 30
Nov 29 19:16:17 dnsmasq[1]: 9721411 10.42.138.0/39928 cached monolith-cache.us-west-2.compute.internal is NXDOMAIN
Nov 29 19:16:17 dnsmasq[1]: 9721412 10.42.138.0/39928 cached monolith-cache.us-west-2.compute.internal is NXDOMAIN
Nov 29 19:16:17 dnsmasq[1]: 9721413 10.42.30.0/46826 reply feed-service.svc.cluster.local is NXDOMAIN
Nov 29 19:16:17 dnsmasq[1]: 9721414 10.42.30.0/46826 reply feed-service.svc.cluster.local is NXDOMAIN
Nov 29 19:16:17 dnsmasq[1]: 9721417 10.42.64.0/52270 reply monolith-cache.cluster.local is NXDOMAIN
Nov 29 19:16:17 dnsmasq[1]: 9721418 10.42.64.0/52270 reply monolith-cache.cluster.local is NXDOMAIN
Nov 29 19:16:17 dnsmasq[1]: 9721420 10.42.86.0/48582 reply monolith-cache.default.svc.cluster.local is NXDOMAIN
Nov 29 19:16:17 dnsmasq[1]: 9721419 10.42.86.0/48582 reply monolith-cache.default.svc.cluster.local is NXDOMAIN
Nov 29 19:16:17 dnsmasq[1]: 9721421 10.42.60.0/32780 reply monolith-cache.svc.cluster.local is NXDOMAIN
Nov 29 19:16:17 dnsmasq[1]: 9721422 10.42.60.0/32780 reply monolith-cache.svc.cluster.local is NXDOMAIN
Nov 29 19:16:17 dnsmasq[1]: 9721423 10.42.60.0/51011 cached monolith-cache.us-west-2.compute.internal is NXDOMAIN
Nov 29 19:16:17 dnsmasq[1]: 9721424 10.42.60.0/51011 cached monolith-cache.us-west-2.compute.internal is NXDOMAIN
Nov 29 19:16:17 dnsmasq[1]: 9721425 10.42.150.0/56602 cached monolith-cache.us-west-2.compute.internal is NXDOMAIN
Nov 29 19:16:17 dnsmasq[1]: 9721426 10.42.150.0/56602 cached monolith-cache.us-west-2.compute.internal is NXDOMAIN
Nov 29 19:16:17 dnsmasq[1]: 9721427 10.42.9.0/42659 cached monolith-cache.us-west-2.compute.internal is NXDOMAIN
Nov 29 19:16:17 dnsmasq[1]: 9721428 10.42.9.0/42659 cached monolith-cache.us-west-2.compute.internal is NXDOMAIN
Nov 29 19:16:17 dnsmasq[1]: 9721429 10.42.7.0/52143 reply feed-service.default.svc.cluster.local is NXDOMAIN
Nov 29 19:16:17 dnsmasq[1]: 9721430 10.42.7.0/52143 reply feed-service.default.svc.cluster.local is NXDOMAIN
Nov 29 19:16:17 dnsmasq[1]: 9721431 10.42.7.0/46174 cached feed-service.us-west-2.compute.internal is NXDOMAIN
Nov 29 19:16:17 dnsmasq[1]: 9721432 10.42.7.0/46174 cached feed-service.us-west-2.compute.internal is NXDOMAIN
Nov 29 19:16:17 dnsmasq[1]: 9721433 10.42.59.0/36538 reply monolith-cache.svc.cluster.local is NXDOMAIN
Nov 29 19:16:17 dnsmasq[1]: 9721434 10.42.59.0/36538 reply monolith-cache.svc.cluster.local is NXDOMAIN
Nov 29 19:16:17 dnsmasq[1]: 9721436 10.42.109.0/38199 reply monolith-cache.default.svc.cluster.local is NXDOMAIN
Nov 29 19:16:17 dnsmasq[1]: 9721435 10.42.109.0/38199 reply monolith-cache.default.svc.cluster.local is NXDOMAIN
Nov 29 19:16:17 dnsmasq[1]: 9721437 10.42.109.0/48251 cached monolith-cache.us-west-2.compute.internal is NXDOMAIN
Nov 29 19:16:17 dnsmasq[1]: 9721438 10.42.109.0/48251 cached monolith-cache.us-west-2.compute.internal is NXDOMAIN
Nov 29 19:16:17 dnsmasq[1]: 9721439 10.42.30.0/38612 reply monolith-cache.svc.cluster.local is NXDOMAIN
Nov 29 19:16:17 dnsmasq[1]: 9721441 10.42.86.0/35070 reply monolith-cache.default.svc.cluster.local is NXDOMAIN
Nov 29 19:16:17 dnsmasq[1]: 9721440 10.42.30.0/38612 reply monolith-cache.svc.cluster.local is NXDOMAIN
Nov 29 19:16:17 dnsmasq[1]: 9721442 10.42.86.0/35070 reply monolith-cache.default.svc.cluster.local is NXDOMAIN
```
Apparently, something in the SOA responses generated by kube-dns prevents dnsmasq from caching negative results.

**Comment by bogdando** [comment]
@fluxrad the `adress=/foo/` doesn't work for me (tried with Centos7). But that http://lists.thekelleys.org.uk/pipermail/dnsmasq-discuss/2011q3/005137.html worked well, so perhaps `local=/foo/` should be used.

**Comment by macb** [comment]
`address=/foo/` worked for me:

When logging is enabled can see lines similar to:
```
dnsmasq[1]: query[A] bar.foo from 172.17.41.0
dnsmasq[1]: config bar.foo is NXDOMAIN
```

**Comment by thockin** [comment]
@bowei fell thru cracks.

**Comment by bowei** [comment]
@vladlosev @bogdando @macb can you file a new issue against the kubernetes/dns regarding the SOA response for negative caching? I'm closing this issue as the topics discussed seem to be diverging from the initial bug report.

**Comment by fluxrad** [comment]
As far as I'm aware, significant volumes of erroneous queries from the `ndots` problem is still an issue. It generates huge load on the DNS infrastructure. Do we at least have a canonical tracking ticket for that? https://github.com/kubernetes/kubernetes/issues/33554 perhaps?

**Comment by bowei** [comment]
Yes -- put references to ndots related issues on #33554
